github/oidc.go

```diff
@@ -15,83 +15,228 @@
-func RequestOIDCToken(audience string) (*OIDCToken, error) {
-	urlKey := os.Getenv(requestURLEnvKey)
-	if urlKey == "" {
-		return nil, errURLEnvKeyEmpty
-	}
+// errToken indicates an error in the format of the token.
+type errToken struct {
+	errors.WrappableError
+}
+
+// errVerify indicates an error in the token verification process.
+type errVerify struct {
+	errors.WrappableError
+}
+
+// OIDCClient is a client for the GitHub OIDC provider.
+type OIDCClient struct {
+	actionsProviderURL string
+
+	// requestURL is the GitHub URL to request a OIDC token.
+	requestURL *url.URL
+
+	// bearerToken is used to request an ID token.
+	bearerToken string
 
-	url := urlKey + "&audience=" + audience
-	req, err := http.NewRequest("GET", url, nil)
+	// verifierFunc is a factory to generate an oidc.IDTokenVerifier for token verification.
+	// This is used for tests.
+	verifierFunc func(context.Context) (*oidc.IDTokenVerifier, error)
+}
+
+// NewOIDCClient returns new GitHub OIDC provider client.
+func NewOIDCClient() (*OIDCClient, error) {
+	requestURL := os.Getenv(requestURLEnvKey)
+	parsedURL, err := url.Parse(requestURL)
 	if err != nil {
-		return nil, fmt.Errorf("creating request: %w", err)
+		return nil, fmt.Errorf("invalid request URL %q: %w", requestURL, err)
+	}
+
+	c := OIDCClient{
+		actionsProviderURL: defaultActionsProviderURL,
+		requestURL:         parsedURL,
+		bearerToken:        os.Getenv(requestTokenEnvKey),
 	}
+	c.verifierFunc = func(ctx context.Context) (*oidc.IDTokenVerifier, error) {
+		provider, err := oidc.NewProvider(ctx, c.actionsProviderURL)
+		if err != nil {
+			return nil, err
+		}
+		return provider.Verifier(&oidc.Config{
+			// NOTE: Disable ClientID check.
+			// ClientID is normally checked to be part of the audience but we
+			// don't use a ClientID when requesting a token.
+			SkipClientIDCheck: true,
+		}), nil
+	}
+	return &c, nil
+}
 
-	req.Header.Add("Authorization", "bearer "+os.Getenv(requestTokenEnvKey))
+func (c *OIDCClient) newRequestURL(audience []string) string {
+	requestURL := *c.requestURL
+	q := requestURL.Query()
+	for _, a := range audience {
+		q.Add("audience", a)
+	}
+	requestURL.RawQuery = q.Encode()
+	return requestURL.String()
+}
+
+func (c *OIDCClient) requestToken(ctx context.Context, audience []string) ([]byte, error) {
+	// Request the token.
+	req, err := http.NewRequest("GET", c.newRequestURL(audience), nil)
+	if err != nil {
+		return nil, errors.Errorf(&errRequestError{}, "creating request: %w", err)
+	}
+	req.Header.Add("Authorization", "bearer "+c.bearerToken)
+	req = req.WithContext(ctx)
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
-		return nil, fmt.Errorf("request: %w", err)
+		return nil, errors.Errorf(&errRequestError{}, "request: %w", err)
 	}
 	defer resp.Body.Close()
 
+	// Read the response.
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Errorf(&errRequestError{}, "reading response: %w", err)
+	}
+	if resp.StatusCode >= 400 {
```

Comment:
Yeah, probably some like 201 would be "ok" but we should probably wouldn't want to support or deal with redirects so returning an error would be fine there.

Reply:
BTW, what gets returned on invalid audience? I get a 4XX error I remember right.

Reply:
I also want to know what's makes for an invalid audience? I couldn't find any info on it. The only issue I could find is that github doesn't like "github.com" to be anywhere in the audience.

Reply:
I don't think it's documented :/

Reply:
> BTW, what gets returned for you on invalid audience? I get a 4XX error I remember right.

I think it was a 3XX for me, because the `>= 400` code did not catch it

Reply:
> Yeah, probably some like 201 would be "ok" but we should probably wouldn't want to support or deal with redirects so returning an error would be fine there.

let's not deal with that. I think the client can handle it automatically... unless we pass a flag to not do so.

Reply:
> > Yeah, probably some like 201 would be "ok" but we should probably wouldn't want to support or deal with redirects so returning an error would be fine there.
> 
> let's not deal with that. I think the client can handle it automatically... unless we pass a flag to not do so.

I'm not sure it does? I'm pretty sure 201 will get returned in the Response just like any other status thought we could just not expect that and return an error, I'm not sure I trust Github to never return a 201.

I guess maybe you mean redirects? I agree we should just not follow them and return an error as we wouldn't expect them, and I don't think we want to deal with the potential security issues.

Reply:
I'm just wondering if GitHub may some days sends redirect because the are slowly migrating to a different server. 
Btw, sorry about saying we should do `!= 200`, I think  `<200 || >=300` would work fine if we don't follow redirects. Wdut?

Reply:
Yeah, `<200 || >=300` SGTM.

I'm sure it's possible they could send back a redirect during migrations or something but I'd imagine that they wouldn't redirect to a different url even in that case, they'd just proxy you you to a different backend or return 503 or something.